Allow packwerk to scan for packages inside of engines

AndrewSwerlick opened the pull request:
## What are you trying to accomplish?

I'm trying to add support to packwerk for more complex repository structures, particularly mono-repos where multiple apps live in the same repo, and share code through engines that are defined in top level folders. Currently, packwerk cannot detect violations where an app is inappropriately using code defined in a package within an engine, because packwerk cannot discover engines that aren't defined below the `Rails.root` of the app. 

## What approach did you choose and why?

To support this, I modified `PackageSet` so it also looks at `Rails.application.railties` to pull a list of engines, and their root directories, and include those in the paths scanned for packages. I also dropped the restrictions on load paths for constant resolution in`ApplicationLoadPaths`. The result is that packwerk can now find packages defined in these engines, and resolve constants in these engines.

With this setup, packwerk still only scans files in the actual app and it's subdirectories for violations, but now it can resolve constants in the engines, and attach them to packages in the engine.

## What should reviewers focus on?

I'm not confident in the approach I took to mocking and testing things to support the new tests I wrote, so particular focus there would be helpful. Also I'm curious if there are implications to the changes to `ApplicationLoadPaths` that I'm not considering. 

## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Non-breaking change (a change that doesn't alter functionality - i.e., code refactor, configs, etc.)

### Additional Release Notes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)

Include any notes here to include in the release description. For example, if you selected "breaking change" above, leave notes on how users can transition to this version.

If no additional notes are necessary, delete this section or leave it unchanged.

## Checklist

- [ ] I have updated the documentation accordingly. _Do we want documentation changes for this, to call it out as a specific use case?_
- [x] I have added tests to cover my changes.
- [x] It is safe to rollback this change.

exterm commented on lib/packwerk/application_load_paths.rb:
```diff
@@ -31,11 +31,9 @@ def extract_application_autoload_paths
 
         all_paths
           .transform_keys { |path| Pathname.new(path).expand_path }
-          .select { |path| path.fnmatch(rails_root_match.to_s) } # path needs to be in application directory
```
I think we introduced this line to keep packwerk from inferring constants to be defined in gems. Not sure it's still necessary as we might now be doing that filtering somewhere else.

AndrewSwerlick replied:
> I think we introduced this line to keep packwerk from inferring constants to be defined in gems

I think that filtering is also happening with the line right below it, but I'll double check testing against our mono repo.

That being said, I do wonder if maybe changing the behavior so packwerk does infer constants defined in gems could be a feature not a bug? I can imagine a use case where a gem author publishes a rails engine, with packages defined inside of it, and they intend for that to message to consumers what the public API of the gem is, vs the internals you shouldn't be relying on. I'm not sure if this use case has been explicitly considered and rejected by packwerk maintainers, or just hadn't been considered yet, and ya'll chose to filter out constants from gems for performance or other reasons.

AndrewSwerlick replied:
> I think that filtering is also happening with the line right below it, but I'll double check testing against our mono repo.

Just closing this loop, testing this function with our own repo, which has both locally defined engines, and engines from ruby gems via bundler, I'm only seeing our locally defined engine paths show show up

AndrewSwerlick commented:
> One thing I did notice is that it appears to increase runtime significantly (~30% increase in runtime). I didn't get to investigating this more thoroughly, but I wonder if it's because there is a lot files to glob out and scan? I'm really surprised it would add 8 seconds like this though, but it appears to be happening consistently across runs.

I have some suspicions as to where this is coming from. I suspect I'm generating effectively overlapping globs, and ruby is scanning some directories multiple times as a result. There's probably a way to filter out those.

> I'd love to dig in a bit deeper on the performance implications. Separately, I realize that this is actually a behavioral change for user, and in theory could be a breaking change. I wonder if (A) to make it non-breaking and (B) unblock you to use this in your application without us digging into the performance implications we can have this be off by default and able to be turned on with a configuration in `packwerk.yml`. I imagine something like `include_external_rails_engines` (there is probably a better name here). I use "external" to distinguish setups that have unbuilt rails engines located within the rails app.

I'm cool with a new configuration key. I'll add it to PR shortly

AlexEvanczuk commented:
> > maybe changing the behavior so packwerk does infer constants defined in gems could be a feature not a bug?
> 
> We'd have to come up with a convention for non-application package _names_. Since package name and path from application root are the same in packwerk, we don't have a convention for the name of a package that lives outside of the application root.
> 
> E.g. the package `components/platform/thingamajig` would _always_ live at the path `$RAILS_ROOT/components/platform/thingamajig`.

I'd love it if packwerk could be used to detect gem dependencies someday. I was excited about this because it could help folks keep their dependencies on other gems down, which ultimately makes packwerk more valuable from the perspective of a tool to help folks extract a system into another service (i.e. you need to explicitly list all your gem dependencies to do so).

That being said as @exterm said we have a handful of questions to answer, but could be cool for us to start with a design brainstorm in a discussion somewhere.

AndrewSwerlick commented:
Just a heads up, the new configuration key is a little trickier than I realized, just because building a package `PackageSet` relies on self methods that don't have access to the `Configuration` object, so I'm ending up with a lot of messy parameter passing.

At this point I think it's probably worth it to refactor these self methods into a new `PackagePaths` class, which takes in a full configuration object and ends up replacing parts of `PackageSet.load_all_from` and `PackageSet.package_paths`. Otherwise this new config key has to get passed down as a dedicated argument to a few different methods that already have 3 arguments. I'm going to plan to do that, and then add support for this key.

AlexEvanczuk commented:
I think that sounds reasonable to me @AndrewSwerlick , though may have some more feedback when I see the diff. Really appreciate you continuing to push through on this, and please let me know if you feel stuck or frustrated with anything and I'd be happy to pair with you again on zoom.

AndrewSwerlick commented:
@AlexEvanczuk I've added the configuration setting. It's made the PR a little more gnarly, so I cleaned up the history so you can see the refactoring changes in one commit, and then the introducing of the new capabilities in the second.

I still need to update the documentation to check off all the boxes.

shageman commented:
I have a suspicion there will be some interaction between our two PRs: https://github.com/Shopify/packwerk/pull/218

AndrewSwerlick commented:
Hey @alexevanczuk, I just wanted to bump this and see if it's still something we can explore getting merged in. I know there's been some conversations in other PRs that may make some of this obsolete, so I'm open to making changes as necessary, but wanted to check in and see where you all stood.

alexevanczuk reviewed (approved):
This looks fine to me @AndrewSwerlick. @rafaelfranca let me know if you have any concerns prior to merging and releasing

alexevanczuk commented:
@AndrewSwerlick Could you try pushing another commit to this branch (and then perhaps reverting it) to see if it kicks off CI?

AndrewSwerlick commented on lib/packwerk/package_paths.rb:
```diff
@@ -0,0 +1,71 @@
+    def engine_paths_to_scan
+      bundle_path_match = Bundler.bundle_path.join("**")
+
+      Rails.application.railties
```
> We don't depend on railties in the gemspec so I'd rather not depend on rails constants here

I'll address this in your second comment below about paths because I think they're kind of related

> I think this also assumes all local engines are packages

That's not exactly accurate. It assumes any local engine might contain packages, but not that the engine itself is a package. The engine doesn't have to contain packages either. It's just adding all the paths to the list of paths that later get scanned for via glob for a `package.yml` file.

rafaelfranca commented:
> Currently, packwerk cannot detect violations where an app is inappropriately using code defined in a package within an engine, because packwerk cannot discover engines that aren't defined below the Rails.root of the app.

I'm not sure exactly why this is needed because what you are describing here is how we use packwerk and it already work for us without modification. Do you have an example of what doesn't work?
